File: binding/src/libqglyphrun.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qglyphrun.html
pub const qglyphrun = struct {

    /// New constructs a new QGlyphRun object.
    pub fn New() QtC.QGlyphRun {
        return qtc.QGlyphRun_new();
    }


    /// New2 constructs a new QGlyphRun object.
    pub fn New2(other: ?*anyopaque) QtC.QGlyphRun {
        return qtc.QGlyphRun_new2(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QGlyphRun_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QGlyphRun_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn RawFont(self: ?*anyopaque, ) QtC.QRawFont {
        return qtc.QGlyphRun_RawFont(@ptrCast(self));
    }

    pub fn SetRawFont(self: ?*anyopaque, rawFont: ?*anyopaque) void {
        qtc.QGlyphRun_SetRawFont(@ptrCast(self), @ptrCast(rawFont));
    }

    pub fn SetRawData(self: ?*anyopaque, glyphIndexArray: *const u32, glyphPositionArray: ?*anyopaque, size: i32) void {
        qtc.QGlyphRun_SetRawData(@ptrCast(self), @ptrCast(glyphIndexArray), @ptrCast(glyphPositionArray), size);
    }

    pub fn GlyphIndexes(self: ?*anyopaque, ) []const u8 {
        return qtc.QGlyphRun_GlyphIndexes(@ptrCast(self));
    }

    pub fn SetGlyphIndexes(self: ?*anyopaque, glyphIndexes: []const u8) void {
        qtc.QGlyphRun_SetGlyphIndexes(@ptrCast(self), @ptrCast(glyphIndexes));
    }

    pub fn Positions(self: ?*anyopaque, ) []const u8 {
        return qtc.QGlyphRun_Positions(@ptrCast(self));
    }

    pub fn SetPositions(self: ?*anyopaque, positions: []const u8) void {
        qtc.QGlyphRun_SetPositions(@ptrCast(self), @ptrCast(positions));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QGlyphRun_Clear(@ptrCast(self));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QGlyphRun_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QGlyphRun_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn SetOverline(self: ?*anyopaque, overline: bool) void {
        qtc.QGlyphRun_SetOverline(@ptrCast(self), overline);
    }

    pub fn Overline(self: ?*anyopaque, ) bool {
        return qtc.QGlyphRun_Overline(@ptrCast(self));
    }

    pub fn SetUnderline(self: ?*anyopaque, underline: bool) void {
        qtc.QGlyphRun_SetUnderline(@ptrCast(self), underline);
    }

    pub fn Underline(self: ?*anyopaque, ) bool {
        return qtc.QGlyphRun_Underline(@ptrCast(self));
    }

    pub fn SetStrikeOut(self: ?*anyopaque, strikeOut: bool) void {
        qtc.QGlyphRun_SetStrikeOut(@ptrCast(self), strikeOut);
    }

    pub fn StrikeOut(self: ?*anyopaque, ) bool {
        return qtc.QGlyphRun_StrikeOut(@ptrCast(self));
    }

    pub fn SetRightToLeft(self: ?*anyopaque, on: bool) void {
        qtc.QGlyphRun_SetRightToLeft(@ptrCast(self), on);
    }

    pub fn IsRightToLeft(self: ?*anyopaque, ) bool {
        return qtc.QGlyphRun_IsRightToLeft(@ptrCast(self));
    }

    pub fn SetFlag(self: ?*anyopaque, flag: i32) void {
        qtc.QGlyphRun_SetFlag(@ptrCast(self), @intCast(flag));
    }

    pub fn SetFlags(self: ?*anyopaque, flags: i32) void {
        qtc.QGlyphRun_SetFlags(@ptrCast(self), @intCast(flags));
    }

    pub fn Flags(self: ?*anyopaque, ) i32 {
        return qtc.QGlyphRun_Flags(@ptrCast(self));
    }

    pub fn SetBoundingRect(self: ?*anyopaque, boundingRect: ?*anyopaque) void {
        qtc.QGlyphRun_SetBoundingRect(@ptrCast(self), @ptrCast(boundingRect));
    }

    pub fn BoundingRect(self: ?*anyopaque, ) QtC.QRectF {
        return qtc.QGlyphRun_BoundingRect(@ptrCast(self));
    }

    pub fn StringIndexes(self: ?*anyopaque, ) []const u8 {
        return qtc.QGlyphRun_StringIndexes(@ptrCast(self));
    }

    pub fn SetStringIndexes(self: ?*anyopaque, stringIndexes: []const u8) void {
        qtc.QGlyphRun_SetStringIndexes(@ptrCast(self), @ptrCast(stringIndexes));
    }

    pub fn SetSourceString(self: ?*anyopaque, sourceString: []const u8) void {
        const sourceString_str = qtc.libqt_string{
    .len = sourceString.len,
    .data = sourceString.ptr,
};
qtc.QGlyphRun_SetSourceString(@ptrCast(self), sourceString_str);
    }

    pub fn SourceString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QGlyphRun_SourceString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qglyphrun.SourceString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QGlyphRun_IsEmpty(@ptrCast(self));
    }

    pub fn SetFlag2(self: ?*anyopaque, flag: i32, enabled: bool) void {
        qtc.QGlyphRun_SetFlag2(@ptrCast(self), @intCast(flag), enabled);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QGlyphRun_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/glyphrun.html#types
pub const enums = struct {
    pub const GlyphRunFlag = enum {
        pub const Overline: i32 = 1;
        pub const Underline: i32 = 2;
        pub const StrikeOut: i32 = 4;
        pub const RightToLeft: i32 = 8;
        pub const SplitLigature: i32 = 16;
    };

};

File: binding/src/libqglyphrun.cpp
#include <QGlyphRun>
#include <QPointF>
#include <QRawFont>
#include <QRectF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qglyphrun.h>
#include "libqglyphrun.h"
#include "libqglyphrun.hxx"

QGlyphRun* QGlyphRun_new() {
	 return new QGlyphRun();
}

QGlyphRun* QGlyphRun_new2(const QGlyphRun* other) {
	 return new QGlyphRun(*other);
}

void QGlyphRun_OperatorAssign(QGlyphRun* self, const QGlyphRun* other) {
	self->operator=(*other);
}

void QGlyphRun_Swap(QGlyphRun* self, QGlyphRun* other) {
	self->swap(*other);
}

QRawFont* QGlyphRun_RawFont(const QGlyphRun* self) {
	return new QRawFont(self->rawFont());
}

void QGlyphRun_SetRawFont(QGlyphRun* self, const QRawFont* rawFont) {
	self->setRawFont(*rawFont);
}

void QGlyphRun_SetRawData(QGlyphRun* self, const unsigned int* glyphIndexArray, const QPointF* glyphPositionArray, int size) {
	self->setRawData(glyphIndexArray, glyphPositionArray, size);
}

libqt_list QGlyphRun_GlyphIndexes(const QGlyphRun* self) {
	auto _ret = self->glyphIndexes();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGlyphRun_SetGlyphIndexes(QGlyphRun* self, const libqt_list glyphIndexes) {
	self->setGlyphIndexes(QList<unsigned int>());
}

libqt_list QGlyphRun_Positions(const QGlyphRun* self) {
	auto _ret = self->positions();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGlyphRun_SetPositions(QGlyphRun* self, const libqt_list positions) {
	self->setPositions(QList<QPointF>());
}

void QGlyphRun_Clear(QGlyphRun* self) {
	self->clear();
}

bool QGlyphRun_OperatorEqual(const QGlyphRun* self, const QGlyphRun* other) {
	return self->operator==(*other);
}

bool QGlyphRun_OperatorNotEqual(const QGlyphRun* self, const QGlyphRun* other) {
	return self->operator!=(*other);
}

void QGlyphRun_SetOverline(QGlyphRun* self, bool overline) {
	self->setOverline(overline);
}

bool QGlyphRun_Overline(const QGlyphRun* self) {
	return self->overline();
}

void QGlyphRun_SetUnderline(QGlyphRun* self, bool underline) {
	self->setUnderline(underline);
}

bool QGlyphRun_Underline(const QGlyphRun* self) {
	return self->underline();
}

void QGlyphRun_SetStrikeOut(QGlyphRun* self, bool strikeOut) {
	self->setStrikeOut(strikeOut);
}

bool QGlyphRun_StrikeOut(const QGlyphRun* self) {
	return self->strikeOut();
}

void QGlyphRun_SetRightToLeft(QGlyphRun* self, bool on) {
	self->setRightToLeft(on);
}

bool QGlyphRun_IsRightToLeft(const QGlyphRun* self) {
	return self->isRightToLeft();
}

void QGlyphRun_SetFlag(QGlyphRun* self, int flag) {
	self->setFlag(static_cast<QGlyphRun::GlyphRunFlag>(flag));
}

void QGlyphRun_SetFlags(QGlyphRun* self, int flags) {
	self->setFlags(static_cast<QFlags<QGlyphRun::GlyphRunFlag>>(flags));
}

int QGlyphRun_Flags(const QGlyphRun* self) {
	return self->flags();
}

void QGlyphRun_SetBoundingRect(QGlyphRun* self, const QRectF* boundingRect) {
	self->setBoundingRect(*boundingRect);
}

QRectF* QGlyphRun_BoundingRect(const QGlyphRun* self) {
	return new QRectF(self->boundingRect());
}

libqt_list QGlyphRun_StringIndexes(const QGlyphRun* self) {
	auto _ret = self->stringIndexes();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGlyphRun_SetStringIndexes(QGlyphRun* self, const libqt_list stringIndexes) {
	self->setStringIndexes(QList<QIntegerForSizeof<std::size_t>::Signed>());
}

void QGlyphRun_SetSourceString(QGlyphRun* self, const libqt_string sourceString) {
	self->setSourceString(QString::fromUtf8(sourceString.data, sourceString.len));
}

libqt_string QGlyphRun_SourceString(const QGlyphRun* self) {
	QString _ret = self->sourceString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QGlyphRun_IsEmpty(const QGlyphRun* self) {
	return self->isEmpty();
}

void QGlyphRun_SetFlag2(QGlyphRun* self, int flag, bool enabled) {
	self->setFlag(static_cast<QGlyphRun::GlyphRunFlag>(flag), enabled);
}

void QGlyphRun_Delete(QGlyphRun* self) {
    delete self;
}

